internal/handlers/tutorials.go

```diff
@@ -60,18 +67,144 @@ func saveTutorialProgress(w http.ResponseWriter, r *http.Request) {
 	ctx := r.Context()
 	log := internalctx.GetLogger(ctx)
 	auth := auth.Authentication.Require(ctx)
-	tutorial := r.PathValue("tutorial")
+	tutorial := types.Tutorial(r.PathValue("tutorial"))
 
 	req, err := JsonBody[api.TutorialProgressRequest](w, r)
 	if err != nil {
 		return
 	}
 
-	if res, err := db.SaveTutorialProgress(ctx, auth.CurrentUserID(), types.Tutorial(tutorial), &req); err != nil {
-		log.Warn("could not save tutorial progress", zap.Error(err))
-		sentry.GetHubFromContext(ctx).CaptureException(err)
-		http.Error(w, err.Error(), http.StatusInternalServerError)
+	_ = db.RunTx(ctx, func(ctx context.Context) error {
+		if tutorial == types.TutorialAgents && req.StepID == "welcome" && req.TaskID == "start" {
+			if dt, err := createSampleAppAndDeployment(ctx); err != nil {
+				log.Warn("could not create sample app and deployment", zap.Error(err))
+				sentry.GetHubFromContext(ctx).CaptureException(err)
+				http.Error(w, "could not create sample app and deployment", http.StatusInternalServerError)
+				return err
+			} else if dt != nil {
+				// TODO save additional data?
+				req.Value = map[string]string{
+					"deploymentTargetId": dt.ID.String(),
+				}
+			}
+		}
+
+		if res, err := db.SaveTutorialProgress(ctx, auth.CurrentUserID(), tutorial, &req); err != nil {
+			log.Warn("could not save tutorial progress", zap.Error(err))
+			sentry.GetHubFromContext(ctx).CaptureException(err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return err
+		} else {
+			RespondJSON(w, res)
+			return nil
+		}
+	})
+}
+
+func createSampleAppAndDeployment(ctx context.Context) (*types.DeploymentTargetWithCreatedBy, error) {
+	auth := auth.Authentication.Require(ctx)
+	var progress *types.TutorialProgress
+	var err error
+	if progress, err = db.GetTutorialProgress(ctx, auth.CurrentUserID(), types.TutorialAgents); err != nil {
+		if !errors.Is(err, apierrors.ErrNotFound) {
+			return nil, fmt.Errorf("failed to get existing tutorial progress: %w", err)
+		}
+	}
+	if progress == nil || !slices.ContainsFunc(progress.Events, func(event types.TutorialProgressEvent) bool {
+		return event.StepID == "welcome" && event.TaskID == "start"
+	}) {
+		if app, err := createHelloDistrApp(ctx); err != nil {
+			return nil, fmt.Errorf("failed to create hello-distr app: %w", err)
+		} else if dt, err := createHelloDistrDeploymentTarget(ctx); err != nil {
+			return nil, fmt.Errorf("failed to create hello-distr deployment target: %w", err)
+		} else if err := createHelloDistrDeploymentAndRevision(ctx, app.Versions[0].ID, dt.ID); err != nil {
+			return nil, fmt.Errorf("failed to deploy hello-distr: %w", err)
+		} else {
+			return dt, nil
+		}
+	}
+	return nil, nil
+}
+
+func createHelloDistrApp(ctx context.Context) (*types.Application, error) {
+	log := internalctx.GetLogger(ctx)
+	auth := auth.Authentication.Require(ctx)
+
+	application := types.Application{
+		Name: "hello-distr",
```

Comment:
Unrelated to this PR but I did the tutorial twice and had two apps with the same name. Would it in general make sense to require unique application names per organization? :thinking: